internal/app/upstream/client.go

```diff
@@ -55,11 +74,179 @@ type Response struct {
+		stream, err = m.edsClient.StreamEndpoints(ctx)
+	default:
+		cancel()
+		m.logger.Error(ctx, "Unsupported Type Url %s", request.GetTypeUrl())
+		return nil, nil, &UnsupportedResourceError{TypeURL: request.GetTypeUrl()}
+	}
+
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+
+	signal := make(chan *version, 1)
+	// The xds protocol https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack
+	// specifies that the first request be empty nonce and empty version.
+	// The origin server will respond with the latest version.
+	signal <- &version{nonce: "", version: ""}
+
+	response := make(chan *v2.DiscoveryResponse)
+
+	go send(ctx, m.logger, cancel, &request, stream, signal, m.callOptions)
+	go recv(ctx, cancel, m.logger, response, stream, signal)
+
+	// We use context cancellation over using a separate channel for signalling stream shutdown.
+	// The reason is cancelling a context tied with the stream is straightforward to signal closure.
+	// Also, the shutdown function could potentially be called more than once by a caller.
+	// Closing channels is not idempotent while cancelling context is idempotent.
+	return response, func() { cancel() }, nil
+}
+
+// It is safe to assume send goroutine will not leak as long as these conditions are true:
+// - SendMsg is performed with timeout.
+// - send is a receiver for signal and exits when signal is closed by the owner.
+// - send also exits on context cancellations.
+func send(
+	ctx context.Context,
+	logger log.Logger,
+	cancelFunc context.CancelFunc,
+	request *v2.DiscoveryRequest,
+	stream grpc.ClientStream,
+	signal chan *version,
+	callOptions CallOptions) {
+	for {
+		select {
+		case sig, ok := <-signal:
+			if !ok {
+				return
+			}
+			request.ResponseNonce = sig.nonce
+			request.VersionInfo = sig.version
+			err := doWithTimeout(ctx, func() error {
+				return stream.SendMsg(request)
+			}, callOptions.Timeout)
+			if err != nil {
+				select {
+				case <-ctx.Done():
+					// Context was cancelled, hence this is not an erroneous scenario.
+				default:
+					logger.Error(ctx, "Error in SendMsg: %s", err.Error())
+				}
+				cancelFunc()
+				return
+			}
+		case <-ctx.Done():
+			_ = stream.CloseSend()
+			return
+		}
+	}
+}
+
+// recv is an infinite loop which blocks on RecvMsg.
+// The only ways to exit the goroutine is by cancelling the context or when an error occurs.
+func recv(
+	ctx context.Context,
+	cancelFunc context.CancelFunc,
+	logger log.Logger,
+	response chan *v2.DiscoveryResponse,
+	stream grpc.ClientStream,
+	signal chan *version) {
+	for {
+		resp := new(v2.DiscoveryResponse)
+		if err := stream.RecvMsg(resp); err != nil {
+			select {
+			case <-ctx.Done():
+				// Context was cancelled, hence this is not an erroneous scenario.
+			default:
+				logger.Error(ctx, "Error in RecvMsg %s", err.Error())
+			}
+			cancelFunc()
+			closeChannels(signal, response)
+			return
+		}
+		select {
+		case <-ctx.Done():
+			closeChannels(signal, response)
```

Comment:
same. break out of the for-loop so that the cleanup method happens in one place.